### ISSUE TITLE
Bug in handling dst transition

### ISSUE THREAD

**Issue by agronholm** [agronholm/apscheduler#115]
Originally reported by: **taranes (Bitbucket: [taranes](http://bitbucket.org/taranes), GitHub: Unknown)**

----------------------------------------

The next fire time of a CronTrigger is wrong if it crosses a dst switch.

Daily trigger: 8 o'clock CE(S)T
But today, the first day with dst this year, the next fire time returned was 9 o'clock CEST.

Here's some code to replicate in ipython
```
#!python
import pytz
from datetime import datetime, timedelta
from apscheduler.triggers.cron import CronTrigger
cet = pytz.timezone('CET')
trigger = CronTrigger(hour=8, timezone=cet)  # once every day at 8 CE(S)T

datetime_cet = cet.localize(datetime(2016, 3, 26, 21))
correct_next_date = cet.localize(datetime(2016, 3, 27, 8))
print 'expected', correct_next_date, '# used cet'
print 'but got ', trigger.get_next_fire_time(None, datetime_cet), '# used cet'

# now in europe/berlin
cet = pytz.timezone('Europe/Berlin')
trigger = CronTrigger(hour=8, timezone=cet)  # once every day at 8 CE(S)T

datetime_cet = cet.localize(datetime(2016, 3, 26, 21))
correct_next_date = cet.localize(datetime(2016, 3, 27, 8))
print 'expected', correct_next_date, '# used europe/berlin'
print 'but got ', trigger.get_next_fire_time(None, datetime_cet), '# used europe/berlin'

# now with a start_date already in the correct dst setting
datetime_cet = cet.localize(datetime(2016, 3, 27, 7))
correct_next_date = cet.localize(datetime(2016, 3, 27, 8))
print 'expected', correct_next_date, '# used europe/berlin after dst change'
print 'and got ', trigger.get_next_fire_time(None, datetime_cet), '# used europe/berlin after dst change'
print 'so the trigger time differs w.r.t. the input timezone. this is unfortunate'

# now some experimenting
print
last_fire_time = cet.localize(datetime(2016, 3, 26, 8))
print last_fire_time, '+ 1 day is', last_fire_time + timedelta(days=1), '# no dst change!'
print last_fire_time, '+ 1 day normalized is', cet.normalize(last_fire_time + timedelta(days=1)), '# bad'
print last_fire_time, 'as naive then + 1 day then to cet is', cet.localize(last_fire_time.replace(tzinfo=None) + timedelta(days=1)), '# that is what I want!'
```
And the output:

```
expected 2016-03-27 08:00:00+02:00 # used cet
but got  2016-03-27 09:00:00+02:00 # used cet
expected 2016-03-27 08:00:00+02:00 # used europe/berlin
but got  2016-03-27 09:00:00+02:00 # used europe/berlin
expected 2016-03-27 08:00:00+02:00 # used europe/berlin after dst change
and got  2016-03-27 08:00:00+02:00 # used europe/berlin after dst change
so the trigger time differs w.r.t. the input timezone. this is unfortunate

2016-03-26 08:00:00+01:00 + 1 day is 2016-03-27 08:00:00+01:00 # no dst change!
2016-03-26 08:00:00+01:00 + 1 day normalized is 2016-03-27 09:00:00+02:00 # bad
2016-03-26 08:00:00+01:00 as naive then + 1 day then to cet is 2016-03-27 08:00:00+02:00 # that is what I want!
```

If I replace lines 117, 118 in _set_field_value of apscheduler.trigger.cron.CronTrigger
```
#!python
        difference = datetime(**values) - dateval.replace(tzinfo=None)
        return self.timezone.normalize(dateval + difference)
```
with
```
#!python
        return self.timezone.localize(datetime(**values))
```
it might do the trick -- but I do not yet unterstand why a difference was constructed in the first place.
I'm not sure if the last two lines in _increment_field_value also play a role in this issue.

Regards and happy easter!

----------------------------------------
- Bitbucket: https://bitbucket.org/agronholm/apscheduler/issue/115

**Comment by backbord** [comment]
I stumbled upon one minor issue:

If I run
```
from apscheduler.triggers.cron import CronTrigger
from datetime import datetime
import pytz
cet = pytz.timezone('CET')
t = CronTrigger(hour='*/2', timezone=cet)  # hit time switch exactly
print 'expected', cet.localize(datetime(2016, 3, 27, 3))
print 'got     ', t.get_next_fire_time(None, cet.localize(datetime(2016, 3, 27, 1, 46)))  # before dst threshold: 2-->3
```
I get the wrong answer:
```
expected 2016-03-27 03:00:00+02:00
got      2016-03-27 02:00:00+01:00
```
This problem stems from `pytz.localize` behaviour (this is strange in my opinion, as I am not explicitly performing arithmetic on a datetime, but merely creating a new instance):
```
In [2]: cet.localize(datetime(2016, 3, 27, 2))
Out[2]: datetime.datetime(2016, 3, 27, 2, 0, tzinfo=<DstTzInfo 'CET' CET+1:00:00 STD>)  # wrong dst setting
In [3]: cet.normalize(cet.localize(datetime(2016, 3, 27, 2)))  # added call to cet.normalize
Out[3]: datetime.datetime(2016, 3, 27, 3, tzinfo=<DstTzInfo 'CET' CEST+2:00:00 DST>)  # right dst setting
```

So I suggest changing https://github.com/agronholm/apscheduler/blob/master/apscheduler/triggers/cron/__init__.py#L135
from
```
return self.timezone.localize(datetime(**values))
```
to
```
return self.timezone.normalize(self.timezone.localize(datetime(**values)))
```

Minor note:
I'm not sure as to why another difference (disregarding dst thresholds) is calculated in `apscheduler.triggers.cron.CronTrigger._increment_field_value`:
  https://github.com/agronholm/apscheduler/blob/master/apscheduler/triggers/cron/__init__.py#L121
So I experimented with it.

Replacing lines 121 and 122 with
  `return self.normalize(self.timezone.localize(datetime(**values))), fieldnum`
similar to the change in method `apscheduler.triggers.cron.CronTrigger._set_field_value` line 135 works fine in my examples (if above suggestion of normalizing is taken into account):
```
from apscheduler.triggers.cron import CronTrigger
from datetime import datetime
import pytz
cet = pytz.timezone('CET')
t = CronTrigger(minute='1-59/30', timezone=cet)
start_date = cet.localize(datetime(2016, 3, 27, 1, 45))  # before dst threshold: 2-->3
expected = cet.localize(datetime(2016, 3, 27, 3, 1))
result = t.get_next_fire_time(None, start_date)
print 'expected', expected
print 'got     ', result

print
t = CronTrigger(hour='*/2', timezone=cet)  # hit time switch exactly
print 'expected', cet.localize(datetime(2016, 3, 27, 3))
print 'got     ', t.get_next_fire_time(None, cet.localize(datetime(2016, 3, 27, 1, 46)))  # before dst threshold: 2-->3
```
Output:
```
expected 2016-03-27 03:01:00+02:00
got      2016-03-27 03:01:00+02:00

expected 2016-03-27 03:00:00+02:00
got      2016-03-27 03:00:00+02:00
```

Hope this helps.

Best regards

**Comment by backbord** [comment]
And there's another issue, from dst=True to dst=False, there is a repeated hour, which is not handled correctly:
```
from apscheduler.triggers.cron import CronTrigger
from datetime import datetime, timedelta
import pytz
utc = pytz.utc
cet = pytz.timezone('CET')
t = CronTrigger(hour='*', minute=30, timezone=cet)
now = utc.localize(datetime(2015, 10, 24, 22, 15))
last_f = cet.normalize(cet.localize(datetime(2015, 10, 25, 1, 30)))
for _ in range(4):
    print 'last_f %s (utc %s), now %s' % (last_f, last_f.astimezone(utc), now)
    last_f = t.get_next_fire_time(last_f, now)
    now += timedelta(hours=1)
```
gives the following output -- note that an hour is skipped:
```
last_f 2015-10-25 01:30:00+02:00 (utc 2015-10-24 23:30:00+00:00), now 2015-10-24 22:15:00+00:00
last_f 2015-10-25 02:30:00+01:00 (utc 2015-10-25 01:30:00+00:00), now 2015-10-25 01:45:00+00:00
last_f 2015-10-25 03:30:00+01:00 (utc 2015-10-25 02:30:00+00:00), now 2015-10-25 02:45:00+00:00
last_f 2015-10-25 04:30:00+01:00 (utc 2015-10-25 03:30:00+00:00), now 2015-10-25 03:45:00+00:00
```
I would have expected 2:30 to be hit twice and the hour in the utc/now columns should not leave out 00:30/23:45:
```
last_f 2015-10-25 01:30:00+02:00 (utc 2015-10-24 23:30:00+00:00), now 2015-10-24 22:15:00+00:00
last_f 2015-10-25 02:30:00+02:00 (utc 2015-10-25 00:30:00+00:00), now 2015-10-25 23:45:00+00:00  # missing
last_f 2015-10-25 02:30:00+01:00 (utc 2015-10-25 01:30:00+00:00), now 2015-10-25 01:45:00+00:00
last_f 2015-10-25 03:30:00+01:00 (utc 2015-10-25 02:30:00+00:00), now 2015-10-25 02:45:00+00:00
last_f 2015-10-25 04:30:00+01:00 (utc 2015-10-25 03:30:00+00:00), now 2015-10-25 03:45:00+00:00
```

The problem appears to stem from ```_increment_field_value``` where the hour value "2" is incremented to "3". This crosses the dst threshold.

Usually, this is coped with by internally working in UTC and UTC only. But with apscheduler CronTrigger's powerful cron-like syntax (e.g. hour="1,5,6,21"), that approach appears quite cumbersome.
In the above example, one might have to increment the hour of the utc datetime "0" to "1" and then translate it to cet.

So, unfortunately, I do not have any elegant and short solution at hand.

But maybe one comes up in the months until the next change from dst=True to dst=False (I guess that's roughly six months in the northern hemisphere, 12 months in the southern hemisphere).

**Comment by backbord** [comment]
How about extending the dst test cases like so:

```
    @pytest.mark.parametrize('trigger_args, start_date, start_date_dst, correct_next_date_dst, correct_next_date', [
        ({'hour': 8}, datetime(2013, 3, 9, 12), False, True, datetime(2013, 3, 10, 8)),
        ({'hour': 8}, datetime(2013, 11, 2, 12), True, False, datetime(2013, 11, 3, 8)),
        ({'hour': '*', 'minute': 30}, datetime(2013, 3, 10, 12), False, False, datetime(2013, 3, 10, 8)),  # repeat hour a
        ({'hour': '*', 'minute': 30}, datetime(2013, 3, 10, 12), False, True, datetime(2013, 3, 10, 8)),  # repeat hour b
        ({'hour': '*'}, datetime(2013, 11, 3, 1), True, False, datetime(2013, 11, 3, 3)),  # skipped hour
        ({'hour': 2, 'minute': 30}, datetime(2013, 11, 2, 12), True, False, datetime(2013, 11, 4, 2, 30)),  
        ({'minute': '*/30'}, datetime(2013, 3, 10, 1, 35), False, True, datetime(2013, 3, 10, 3)),
        ({'minute': '*/30'}, datetime(2013, 11, 3, 1, 35), True, False, datetime(2013, 11, 3, 1))
    ], ids=['absolute_spring', 'absolute_autumn',
            'repeat_hour_a', 'repeat_hour_b',
            'absolute_skipped_hour', 'interval_skipped_hour',
            'interval_spring', 'interval_autumn'])
    def test_dst_change(self, trigger_args, start_date, start_date_dst, correct_next_date):
        """
        Making sure that CronTrigger works correctly when crossing the DST switch threshold.
        Note that you should explicitly compare datetimes as strings to avoid the internal datetime
        comparison which would test for equality in the UTC timezone.

        """
        utc = pytz.utc
        timezone = pytz.timezone('US/Eastern')
        trigger = CronTrigger(timezone=timezone, **trigger_args)
        start_date = timezone.localize(start_date, is_dst=start_date_dst)
        correct_next_date = timezone.localize(correct_next_date, is_dst=correct_next_date_dst)
        assert str(trigger.get_next_fire_time(None, start_date)) == str(correct_next_date)
```

(I currently do not have my full development suite available, so the test code is untested -- however, I'm relatively certain that at least repeat_hour_* and *_skipped_hour will fail.)

I sadly suggest to rebuild some of the inner workings to calculate times in utc, convert them to the desired timezones and check if the desired timestamps can be matched (that's what I did before using apscheduler [and thinking: timezone support? great! someone other than me took care of all my worries regarding DSTs. :-)]).
Timezones and DST's cause headaches.

**Comment by agronholm** [comment]
Possibly still broken.

**Comment by backbord** [comment]
As you wrote in #116, the behaviour shown in my tests above agrees with your documentation (see https://github.com/agronholm/apscheduler/issues/116#issuecomment-202750741).

I believe it's fixed. :-)

**Comment by agronholm** [comment]
I've released 3.0.6 with the fix.
